fs-index/src/index.rs

```diff
@@ -1,1120 +1,501 @@
+            path_to_resource,
         })
     }
 
-    #[test]
-    fn should_not_index_1_empty_directory() {
-        run_test_and_clean_up(|path| {
-            create_dir_at(path.clone());
-
-            let actual: ResourceIndex<Crc32> =
-                ResourceIndex::build(path.clone());
+    /// Update the index with the latest information from the file system
+    pub fn update_all(&mut self) -> Result<IndexUpdate<Id>> {
+        log::debug!("Updating index at root path: {:?}", self.root);
+
+        let mut added = Vec::new();
+        let mut modified = Vec::new();
+        let mut removed = Vec::new();
+
+        let new_index = ResourceIndex::build(&self.root)?;
+
+        // Compare the new index with the old index
+        let current_resources = self.resources();
+        let new_resources = new_index.resources();
+        for resource in new_resources.clone() {
+            // If the resource is in the old index, check if it has been
+            // modified
+            if let Some(current_resource) =
+                self.get_resource_by_path(&resource.path)
+            {
+                if current_resource != &resource {
+                    modified.push(resource.clone());
+                }
+            }
+            // If the resource is not in the old index, it has been added
+            else {
+                added.push(resource.clone());
+            }
+        }
+        for resource in current_resources {
+            // If the resource is not in the new index, it has been removed
+            if !new_resources.contains(&resource) {
+                removed.push(resource.clone());
+            }
+        }
 
-            assert_eq!(actual.root, path.clone());
-            assert_eq!(actual.path2id.len(), 0);
-            assert_eq!(actual.id2path.len(), 0);
-            assert_eq!(actual.collisions.len(), 0);
+        // Update the index with the new index and return the result
+        *self = new_index;
+        Ok(IndexUpdate {
+            added,
+            modified,
+            removed,
         })
     }
 
-    #[test]
-    fn discover_paths_should_not_walk_on_invalid_path() {
-        run_test_and_clean_up(|path| {
-            let mut missing_path = path.clone();
-            missing_path.push("missing/directory");
-            let actual = discover_paths(missing_path);
-            assert_eq!(actual.len(), 0);
-        })
-    }
+    /// Track the addition of a newly added file to the resource index.
+    ///
+    /// This method checks if the file exists in the file system.
+    ///
+    /// # Arguments
+    /// * `relative_path` - The path of the file to be added (relative to the
+    ///   root path of the index).
+    ///
+    /// # Returns
+    /// Returns `Ok(resource)` if the file was successfully added to the index.
+    ///
+    /// # Errors
+    /// - If the file does not exist in the file system.
+    /// - If there was an error calculating the checksum of the file.
```

Comment:
The older comment:
```
 354     /// The caller must ensure that:
 355     /// * the index is up-to-date except this single path
 356     /// * the path hasn't been indexed before
 357     ///
 358     /// Should only be used if reactive updates are not possible.
```

However, now I think we should care about id, not path. The use-case is an app that created resource, and it's very unlikely to have this resource indexed already (user created new image for instance). Then we can just add the resource to the index, skipping the whole index update.

So, the caller must ensure that:
1. The index is up-to-date except this single id.
2. The id hasn't been indexed before.

All `track` functions are intended to be used by app developers for optimized index manipulations, but they should understand the circumstances.